File: ch7/ch7_main.c
#include <stdio.h>
#include <stdlib.h>
#include "uthash.h"

typedef unsigned long ulong;


// LINKED LIST

typedef struct LinkedListNode {
    ulong telephone;
    ulong contact_time;
    struct LinkedListNode* next;
} LinkedListNode;

int LL_add(LinkedListNode* list, ulong tel, ulong c_time) {
    LinkedListNode* current_node = list;
    while (current_node != NULL) {
        if (current_node->telephone == tel)
            return 0; // Item already in list
        current_node = current_node->next;
    }

    // Item not found in list, add it
    LinkedListNode* new_node = malloc( sizeof(LinkedListNode) );
    new_node->telephone = tel;
    new_node->contact_time = c_time;
    new_node->next = NULL;
    current_node->next = new_node;
    return 1;
}

LinkedListNode* LL_add_ordered(LinkedListNode* list, ulong tel, ulong c_time) {
    LinkedListNode* new_node = malloc( sizeof(LinkedListNode) );
    new_node->telephone = tel;
    new_node->contact_time = c_time;    

    if (c_time == list->contact_time) { // Base case only the first node added after the initial has the same time
        list->next = new_node;
        new_node->next = NULL;
        return list;
    }

    if (c_time < list->contact_time) {
        new_node->next = list;
        return new_node;
    }

    if (list->next != NULL) {
        list->next = new_node;
        new_node->next = NULL;
        return list;
    }

    LinkedListNode* previous_node = list;
    LinkedListNode* current_node = list->next;
    while (current_node->next != NULL && current_node->contact_time < c_time) {
        current_node = current_node->next;
    }
    previous_node->next = new_node;
    new_node->next = current_node;
    return list;
}

int LL_contains(LinkedListNode* list, ulong tel) {
    if (list == NULL) return 0; // return false
    while(list != NULL) {
        if (list->telephone == tel) return 1; // return true
        list = list->next;
    }
    return 0; // return false
}


// CONTACTS HASH

struct ContactsHash* contacts_hash = NULL;

typedef struct ContactsHash {
    UT_hash_handle hh;
    ulong telephone;
    LinkedListNode* contacts_list;
} ContactsHash;

int HT_add(ulong caller, ulong answer, ulong c_time) {
    ContactsHash *s;

    HASH_FIND_INT(contacts_hash, &caller, s);
    if (s == NULL) {
        //printf("Adding into HT %lu -> %lu @ %lu\n", caller, answer, c_time);
        
        s = (ContactsHash*) malloc( sizeof(ContactsHash) );
        s->telephone = caller;
        s->contacts_list = malloc( sizeof(LinkedListNode) );
            s->contacts_list->telephone = answer;
            s->contacts_list->contact_time = c_time;
            s->contacts_list->next = NULL;

        HASH_ADD_INT(contacts_hash, telephone, s );  /* telephone: name of key field */
        return 1; // Item inserted
    }

    //printf("Adding into LL %lu -> %lu @ %lu\n", caller, answer, c_time);
    return LL_add(s->contacts_list, answer, c_time);
}

ContactsHash* HT_find(ulong caller) {
    ContactsHash* s;
    HASH_FIND_INT(contacts_hash, &caller, s);
    return s;
}





int main(int argc, char* argv)
{
    FILE* ifp = fopen("phone_call.log", "r");
    ulong first, second;

    int count = 0;
    while ( count < 2 && !feof(ifp) ) {
        fscanf(ifp, "%lu %lu\n", &first, &second);
        
        if ( !HT_add(first, second, count) ) {
            printf("[ERROR] Adding %lu -> %lu @ %lu", first, second, count);
            exit(1);
        }
        if ( !HT_add(second, first, count) ) {
            printf("[ERROR] Adding %lu -> %lu @ %lu", second, first, count);
            exit(1);
        }

        count++;
    }

    //printf("Hash contacts: %d\n", HASH_COUNT(contacts_hash));

    ulong solution = 0;
    // Get terrorist numbers from input
    ulong terr1, terr2;
    scanf("%lu\n%lu\n", &terr1, &terr2);

    LinkedListNode* pending_tels = malloc( sizeof(LinkedListNode) ); // Telephones waiting for being investigated
    pending_tels->telephone = terr1;
    pending_tels->contact_time = 0;
    pending_tels->next = NULL;

    LinkedListNode* evaluated_tels = malloc( sizeof(LinkedListNode) ); // Telephones already investigated
    evaluated_tels->telephone = terr1;
    evaluated_tels->next = NULL;

    while (pending_tels != NULL) {
        
        // Second terrorist reached
        if (pending_tels->telephone == terr2) {
            if (pending_tels->contact_time > solution)
                solution = pending_tels->contact_time;
            printf("Connected at %lu\n", solution);
            exit(0);
        }

        ContactsHash* current_hash = HT_find(pending_tels->telephone);
        LinkedListNode* c_list = current_hash->contacts_list;
        while (c_list != NULL) {
            if ( !LL_contains(evaluated_tels, c_list->telephone) ) {
                pending_tels = LL_add_ordered(pending_tels, c_list->telephone, c_list->contact_time);
            }
            c_list = c_list->next;
        }

        LL_add(evaluated_tels, pending_tels->telephone, 0);
        // Free node evaluated, and jump to the next one
        LinkedListNode* temp = pending_tels;
        pending_tels =pending_tels->next;
        free(temp);
    }

    printf("Not connected");
}

File: ch7/ch7_main.rb
#! /usr/bin/env ruby
require 'json'

# Read phone call log and build a data structure of contacts
call_log = File.open 'phone_call.log', 'r'
calls = call_log.readlines

contacts_hash = Hash.new{ |hash, key| hash[key] = Hash.new } # key = caller number, data = hash {key = called, data = contact_time}

calls.each_with_index do |call_data, index|
  tel_numbers = call_data.chomp.split ' '
  tel_numbers.map! { |n| n.to_i }

  tel_numbers.each_with_index do |tel, j|
    contacts_hash[tel][tel_numbers[(j + 1) % 2]] = index unless contacts_hash[tel][tel_numbers[(j + 1) % 2]]
  end
end

# Get terrorist phone numbers from stdin
terr1 = gets.chomp.to_i
terr2 = gets.chomp.to_i

contact_time = -1 # Time of contact between the terrorists

pending_tels = [] # Ordered list of telephones waiting for being evaluated on the search for contact
pending_tels[0] = [terr1, -1]
index_first_pending = 0
index_last_pending = 0

evaluated_tels = [] # To avoid repeating scan of the same tel

while index_first_pending <= index_last_pending

  while not (tel = pending_tels[index_first_pending]) && index_first_pending <= index_last_pending # Skip empty spaces in the array
    index_first_pending += 1
  end
  pending_tels[index_first_pending] = nil
  
  if tel[0] == terr2
    contact_time = tel[1] unless tel[1] < contact_time
    puts 'Connected at ' << contact_time.to_s
    exit 0
  end

  contacts_hash[tel[0]].each_pair do |key, value|
    pending_tels[value] = [key, value] unless evaluated_tels.include? key
  
    # Update indexes
    index_first_pending = value if value < index_first_pending
    index_last_pending = value if value > index_last_pending
  end

  evaluated_tels << tel[0]
end

puts 'Not connected'
